types/src/validator_verifier.rs

```diff
@@ -159,75 +184,69 @@ impl<PublicKey: VerifyingKey> ValidatorVerifier<PublicKey> {
         Ok(())
     }
 
-    /// Ensure there are at least quorum_size and not more than maximum expected signatures.
-    fn check_num_of_signatures<T>(
+    /// Ensure there is at least quorum_voting_power in the provided signatures and there
+    /// are only known authors. According to the threshold verification policy,
+    /// invalid public keys are not allowed.
+    pub fn check_voting_power_and_signatures<'a>(
         &self,
-        aggregated_signature: &HashMap<AccountAddress, T>,
-    ) -> std::result::Result<(), VerifyError>
-    where
-        T: Into<PublicKey::SignatureMaterial> + Clone,
-    {
-        let num_of_signatures = aggregated_signature.len();
-        if num_of_signatures < self.quorum_size {
-            return Err(VerifyError::TooFewSignatures {
-                num_of_signatures,
-                quorum_size: self.quorum_size,
-            });
+        aggregated_signature: impl Iterator<Item = &'a AccountAddress>,
+    ) -> std::result::Result<(), VerifyError> {
+        let mut aggregated_voting_power = 0;
+        for account_address in aggregated_signature {
+            match self.get_voting_power(&account_address) {
+                Some(voting_power) => aggregated_voting_power += voting_power,
+                None => {
+                    return Err(VerifyError::UnknownAuthor);
+                }
+            }
         }
-        if num_of_signatures > self.len() {
-            return Err(VerifyError::TooManySignatures {
-                num_of_signatures,
-                num_of_authors: self.len(),
+
+        if aggregated_voting_power < self.quorum_voting_power {
+            return Err(VerifyError::TooLittleVotingPower {
+                voting_power: aggregated_voting_power,
+                quorum_voting_power: self.quorum_voting_power,
             });
         }
         Ok(())
     }
 
-    /// Ensure there are only known authors. According to the threshold verification policy,
-    /// invalid public keys are not allowed.
-    fn check_keys<T>(
-        &self,
-        aggregated_signature: &HashMap<AccountAddress, T>,
-    ) -> std::result::Result<(), VerifyError>
-    where
-        T: Into<PublicKey::SignatureMaterial> + Clone,
-    {
-        if aggregated_signature
-            .keys()
-            .all(|author| self.author_to_public_keys.get(&author).is_some())
-        {
-            Ok(())
-        } else {
-            Err(VerifyError::UnknownAuthor)
+    /// Returns the public key for this address.
+    pub fn get_public_key(&self, author: &AccountAddress) -> Option<PublicKey> {
+        match self.address_to_validator_info.get(&author) {
+            Some(validator_info) => Some(validator_info.public_key.clone()),
+            None => None,
         }
```

Comment:
I believe you can just `self.address_to_validator_info.get(&author).map(|vi| vi.public_key.clone())`